src/service/viewport/viewport-impl.js

```diff
@@ -396,42 +400,30 @@ export class Viewport {
   }
 
   /**
-   * Returns a promise that resolves with latest rect of the element within the document.
+   * Returns the clientRect of the element.
+   * If root is viewport: return the element clientRect
+   * If root is an iframe: return the element clientRect if element intersects with root
+   * TODO(@zhouyx): We should consider return null if root doesn't intersect with viewportRect as well.
    * @param {!Element} el
-   * @return {!Promise<!../../layout-rect.LayoutRectDef>}
+   * @return {!Promise<?../../layout-rect.LayoutRectDef>}
    */
-  getLayoutRectAsync(el) {
-    const scrollLeft = this.getScrollLeft();
-    const scrollTop = this.getScrollTop();
-
-    // Go up the window hierarchy through friendly iframes.
-    const frameElement = getParentWindowFrameElement(el, this.ampdoc.win);
-    if (frameElement) {
-      const bPromise = this.binding_.getLayoutRectAsync(el, 0, 0);
-      const cPromise = this.binding_.getLayoutRectAsync(
-          frameElement, scrollLeft, scrollTop);
-      return Promise.all([bPromise, cPromise]).then(values => {
-        const b = values[0];
-        const c = values[1];
-        return layoutRectLtwh(Math.round(b.left + c.left),
-            Math.round(b.top + c.top),
-            Math.round(b.width),
-            Math.round(b.height));
-      });
-    }
-
-    return this.binding_.getLayoutRectAsync(el, scrollLeft, scrollTop);
-  }
+  getClientRectAsync(el) {
+    const local = this.vsync_.measurePromise(() => {
+      return el./*OK*/getBoundingClientRect();
+    });
+    const root = this.binding_.getRootClientRectAsync();
 
-  /**
-   * Returns a promise that resolve with latest rect of the element within the viewport.
-   * @param {!Element} el
-   * @return {!Promise<?../../layout-rect.LayoutRectDef>}}
-   */
-  getBoundingRectAsync(el) {
-    const viewportRect = this.getRect();
-    return this.getLayoutRectAsync(el).then(rect => {
-      return moveLayoutRect(rect, -viewportRect.left, -viewportRect.top);
+    return Promise.all([local, root]).then(values => {
+      const l = values[0];
+      const r = values[1];
+      if (!r) {
+        return l;
+      }
+      const clientRect = moveLayoutRect(l, r.left, r.top);
+      if (!rectIntersection(clientRect, r)) {
```

Comment:
As discussed, let's remove `null` for now. But let's document that occlusion is possible.